Add RC option for quadplane manual forward throttle

kd0aij opened the pull request:
To test, set RCx_OPTION to 209 where x is the RC input channel controlling forward throttle
and set parameter Q_FWD_THR_MAX to the maximum value in percent. For the Convergence tilt vectored tricopter frame, 25% is a reasonable maximum.

IamPete1 commented on libraries/RC_Channel/RC_Channel.h:
```diff
@@ -181,6 +181,7 @@ class RC_Channel {
         DISARM =              81, // disarm vehicle
         Q_ASSIST =            82, // disable, enable and force Q assist
         ZIGZAG_Auto =         83, // zigzag auto switch
+        FWD_THR =             84, // VTOL manual forward throttle
```
I have been adding manual control analog inputs at the end to line up with where the four main ones will end up when we finally kill RCMAP. So this would be 209

kd0aij replied:
thanks, changed to 209

IamPete1 commented on ArduPlane/quadplane.h:
```diff
@@ -335,6 +335,11 @@ class QuadPlane
     AP_Float throttle_expo;
 
+    // manual forward throttle input
+    AP_Int8  fwd_thr_chan;
```
I think this is not needed anymore

kd0aij replied:
good catch, removed

IamPete1 commented on ArduPlane/tiltrotor.cpp:
```diff
@@ -92,29 +92,48 @@ void QuadPlane::tiltrotor_continuous_update(void)
+        (plane.control_mode == &plane.mode_qacro ||
+         plane.control_mode == &plane.mode_qstabilize ||
+         plane.control_mode == &plane.mode_qhover)) {
+        if (rc_fwd_thr_ch == nullptr) {
```
I think is better to just re-get the pointer each time, then it works straight away without a reboot

kd0aij replied:
I moved the find call from init to the one_second_loop, so we're not doing a linear search in the fast loop, but can still change it without rebooting.

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -2910,24 +2920,50 @@ void QuadPlane::Log_Write_QControl_Tuning()
+int8_t QuadPlane::forward_throttle_pct(bool tiltrotor)
+{
+    // if not a tiltrotor, disable forward motor:
+    if (!tiltrotor) {
```
I think we should be able to make this work for all quad planes

kd0aij replied:
That comment may be confusing... manual forward throttle does work for standard quadplanes (SITL tested on the RF8 BigStik quadplane), just not when disarmed.

The change in behavior here is only that tiltrotor motors can tilt forward when disarmed.  This is a good thing because you don't want them to suddenly tilt forward on arming since some planes (like the Convergence) have no clearance to the ground.  
This method is only called with tiltrotor==true from tiltrotor_continuous_update, and that method only affects tilt servos, not throttles.

For a non-tiltrotor, we must return zero here when disarmed or spooled down, otherwise props could spin when they shouldn't.

Hwurzburg commented:
just a nit...but wouldn't the name Q_FWD_MANTHR_MAX (still fits I think) or Q_FWD_RC_THR_MAX be more descriptive and not lead one to think this might limit throttle when using Q_VFWD_GAIN?

IamPete1 commented on ArduPlane/quadplane.h:
```diff
@@ -101,7 +101,7 @@ class QuadPlane
 
     // return desired forward throttle percentage
-    int8_t forward_throttle_pct(void);        
+    int8_t forward_throttle_pct(bool tiltrotor);        
```
you could default this to false and maybe make the bool return_disarmed or something. tilltrotor only makes sense from the context of the function calls but not within the function itself

kd0aij replied:
saved a few lines of code and (I think) eliminated confusion by moving the arming check up to the caller (this is only called from 2 places)

kd0aij commented:
@Hwurzburg Q_FWD_MANTHR_MAX is 16 characters, which is exactly the max allowed length 👍

IamPete1 commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -92,7 +92,7 @@ const AP_Param::GroupInfo RC_Channel::var_info[] = {
     // @DisplayName: RC input option
     // @Description: Function assigned to this RC channel
-    // @Values{Copter}: 0:Do Nothing, 2:Flip, 3:Simple Mode, 4:RTL, 5:Save Trim, 7:Save WP, 9:Camera Trigger, 10:RangeFinder, 11:Fence, 13:Super Simple Mode, 14:Acro Trainer, 15:Sprayer, 16:Auto, 17:AutoTune, 18:Land, 19:Gripper, 21:Parachute Enable, 22:Parachute Release, 23:Parachute 3pos, 24:Auto Mission Reset, 25:AttCon Feed Forward, 26:AttCon Accel Limits, 27:Retract Mount, 28:Relay On/Off, 2 ...
+    // @Values{Copter}: 0:Do Nothing, 2:Flip, 3:Simple Mode, 4:RTL, 5:Save Trim, 7:Save WP, 9:Camera Trigger, 10:RangeFinder, 11:Fence, 13:Super Simple Mode, 14:Acro Trainer, 15:Sprayer, 16:Auto, 17:AutoTune, 18:Land, 19:Gripper, 21:Parachute Enable, 22:Parachute Release, 23:Parachute 3pos, 24:Auto Mission Reset, 25:AttCon Feed Forward, 26:AttCon Accel Limits, 27:Retract Mount, 28:Relay On/Off, 2 ...
```
I think you meant to add to plane rather than copter

kd0aij replied:
I"m going to need a bigger font :)
Also changed to 209 and fixed the order of the Plane options

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2910,24 +2917,41 @@ void QuadPlane::Log_Write_QControl_Tuning()
-    if (!in_vtol_mode() ||
-        !motors->armed() ||
-        vel_forward.gain <= 0 ||
+    if (plane.control_mode == &plane.mode_qacro ||
```
this change removes some of the protections we had, specifically the ground idle check, the motors armed check and the vtol mode check

kd0aij replied:
As the commit message says, the intent is to move the checks to the caller
unless I screwed up the logic...

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2910,24 +2917,41 @@ void QuadPlane::Log_Write_QControl_Tuning()
+    }
+
+    /*
+      in qautotune mode or modes without a velocity controller
```
comment wrong

kd0aij replied:
looks right to me; what is the issue?

tridge commented on ArduPlane/servos.cpp:
```diff
@@ -504,9 +504,14 @@ void Plane::set_servos_controlled(void)
-            constrain_int16(quadplane.forward_throttle_pct(), min_throttle, max_throttle));
+        int16_t fwd_thr = 0;
+        // if armed and not spooled down ask quadplane code for forward throttle
+        if (quadplane.motors->armed() &&
```
why is this moved here?

kd0aij replied:
because this is the only place where the check is needed

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -484,6 +484,13 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
     // @User: Standard
     AP_GROUPINFO("TAILSIT_GSCMIN", 18, QuadPlane, tailsitter.gain_scaling_min, 0.4),
 
+    // @Param: FWD_MANTHR_MAX
```
the param name and description don't make it clear that this is related to tilting rotors

kd0aij replied:
it applies to all quadplanes

tridge replied:
ok. i think the description should reference the RCn_OPTION being set to create a fwd throttle option

kd0aij replied:
expand param desc and add arming check

tridge commented:
needs an arming check for zero throttle

kd0aij commented:
@tridge @WickedShell  I discovered when adding the pre-arm check that throttle is ignored by default (RC_OPTIONS defaults to zero which means arming_check_throttle returns false)
Is this the intended behavior?  seems backwards

WickedShell commented:
@kd0aij yeah... I was unable to carry the opinion we should check it by default on the dev call, so it was an opt in arming check, while roll/pitch/yaw are opt out....

kd0aij commented:
@tridge squashed and retested in SITL
new throttle pre-arm check works if enabled in RC_OPTIONS (non-default)
Ready to merge unless you'd prefer the arming check active by default.

kd0aij commented:
@peterbarker Which "comments from Pete are unresolved"?
I thought I had addressed all of those from @IamPete1

peterbarker commented:
@kd0aij I usually hit "Resolve Conversation" in the github interface if the fix has been done.

peterbarker commented on ArduPlane/quadplane.cpp:
```diff
@@ -2910,24 +2917,41 @@ void QuadPlane::Log_Write_QControl_Tuning()
+    }
+
+    /*
+      in qautotune mode or modes without a velocity controller
```
```suggestion
      Vehicle is in in a mode without a velocity controller
```

... on the assumption that qautotune *does* have a velocity controller.

kd0aij replied:
this zeros out the forward throttle if in qautotune or if the relevant gain is zero
We want zero airspeed in qautotune, so the forward motor stays shut down.

kd0aij commented:
@peterbarker wow, there were a lot of those buttons :)
Thanks for looking at this and #14003...

Hwurzburg commented:
the throttle arming check should pass if is_flying is true.....

kd0aij commented:
After discussion with @Hwurzburg and @tridge on pre-arm throttle checks, I have changed the check to a threshold of 5% 
and changed it to be on by default: the option is now to skip the check instead.  The check is also skipped if AP_Notify::flags.flying is true to allow for in-air rearming.
*** Note that the RPY controls neutral check is not skipped when flying; perhaps it should also be skipped when already flying?

kd0aij commented:
@andyp1per This is repeatedly failing step 4 of the fly_glyro_fft test, but I can't see why this PR would break it...

andyp1per commented:
@kd0aij - it failed to takeoff, that's why the test failed - which sounds related?

andyp1per commented:
The "wanted 4" thing is because it failed to become armable

kd0aij commented:
@andyp1per Thanks... how did you determine that? Do you have to run the test locally to see the buildlogs?

andyp1per commented:
I'm expert at knowing what the failures look like ;)
